metrics/certificaterequest.go

```diff
@@ -0,0 +1,102 @@
+)
+
+var readyConditionStatuses = [...]cmmeta.ConditionStatus{
+	cmmeta.ConditionTrue,
+	cmmeta.ConditionFalse,
+	cmmeta.ConditionUnknown,
+}
+
+// UpdateCertificateRequest will update the given CertificateRequest's metrics for its expiry, renewal, and status condition.
+func (m *Metrics) UpdateCertificateRequest(cr *cmapi.CertificateRequest, exp, renewal time.Time) {
+	m.updateCertificateRequestExpiryAndRenewalTime(cr, exp, renewal)
+	m.updateCertificateRequestStatus(cr)
+}
+
+// updateCertificateRequestExpiryAndRenewalTime updates the expiry and renewal time of a certificate request
+func (m *Metrics) updateCertificateRequestExpiryAndRenewalTime(cr *cmapi.CertificateRequest, exp, renewal time.Time) {
+	expiryTime := 0.0
+	if !exp.IsZero() {
+		expiryTime = float64(exp.Unix())
+	}
+	m.certificateRequestExpiryTimeSeconds.With(prometheus.Labels{
+		"name":         cr.Name,
+		"namespace":    cr.Namespace,
+		"issuer_name":  cr.Spec.IssuerRef.Name,
+		"issuer_kind":  cr.Spec.IssuerRef.Kind,
+		"issuer_group": cr.Spec.IssuerRef.Group}).Set(expiryTime)
+
+	renewalTime := 0.0
+	if !renewal.IsZero() {
+		renewalTime = float64(renewal.Unix())
+	}
+	m.certificateRequestRenewalTimeSeconds.With(prometheus.Labels{
+		"name":         cr.Name,
+		"namespace":    cr.Namespace,
+		"issuer_name":  cr.Spec.IssuerRef.Name,
+		"issuer_kind":  cr.Spec.IssuerRef.Kind,
+		"issuer_group": cr.Spec.IssuerRef.Group}).Set(renewalTime)
+}
+
+// updateCertificateRequestStatus will update the metric for that Certificate Request
+func (m *Metrics) updateCertificateRequestStatus(cr *cmapi.CertificateRequest) {
+	for _, c := range cr.Status.Conditions {
+		if c.Type == cmapi.CertificateRequestConditionReady {
+			m.updateCertificateRequestReadyStatus(cr, c.Status)
+			return
+		}
+	}
+
+	// If no status condition set yet, set to Unknown
+	m.updateCertificateRequestReadyStatus(cr, cmmeta.ConditionUnknown)
+}
+
+func (m *Metrics) updateCertificateRequestReadyStatus(cr *cmapi.CertificateRequest, current cmmeta.ConditionStatus) {
+	for _, condition := range readyConditionStatuses {
+		value := 0.0
+
+		if current == condition {
+			value = 1.0
+		}
+
+		m.certificateRequestReadyStatus.With(prometheus.Labels{
+			"name":         cr.Name,
+			"namespace":    cr.Namespace,
+			"condition":    string(condition),
+			"issuer_name":  cr.Spec.IssuerRef.Name,
+			"issuer_kind":  cr.Spec.IssuerRef.Kind,
+			"issuer_group": cr.Spec.IssuerRef.Group,
+		}).Set(value)
+	}
+}
+
+// RemoveCertificateRequest will delete the CertificateRequest metrics from continuing to be exposed.
+func (m *Metrics) RemoveCertificateRequest(name, namespace string) {
+	m.certificateRequestExpiryTimeSeconds.DeletePartialMatch(prometheus.Labels{"name": name, "namespace": namespace})
+	m.certificateRequestRenewalTimeSeconds.DeletePartialMatch(prometheus.Labels{"name": name, "namespace": namespace})
+	m.certificateRequestReadyStatus.DeletePartialMatch(prometheus.Labels{"name": name, "namespace": namespace})
+}
```

Comment:
I think we should use a Prometheus Collector for metrics derived from API resources. This will prevent bugs like the one fixed by https://github.com/cert-manager/cert-manager/pull/7856. API resources should be cached in all controllers, so just querying the API, which should hit cache, is not a problem.

Reply:
I kind of agree with erik on this one.. we should migrate all of our metrics related to API to collectors.. this makes it easier to also cleanup the metrics when the resources are gone